File: shared/src/commonTest/kotlin/com/ttcoachai/shared/drill/ForwardStrokeFilterTest.kt
package com.ttcoachai.shared.drill

import com.ttcoachai.shared.models.Coco17
import com.ttcoachai.shared.models.Handedness
import com.ttcoachai.shared.models.Keypoint2D
import com.ttcoachai.shared.models.PoseFrame2D
import com.ttcoachai.shared.models.Stroke2D
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ForwardStrokeFilterTest {

    /** Player faces +x: nose ahead of shoulder-mid. Wrist x per frame from [wristXs]. */
    private fun frames(wristXs: List<Float>, noseX: Float = 0.55f): List<PoseFrame2D> =
        wristXs.mapIndexed { i, wx ->
            val kp = MutableList(17) { Keypoint2D(0.5f, 0.5f, 1f) }
            kp[Coco17.NOSE] = Keypoint2D(noseX, 0.15f, 1f)
            kp[Coco17.LEFT_SHOULDER] = Keypoint2D(0.49f, 0.30f, 1f)
            kp[Coco17.RIGHT_SHOULDER] = Keypoint2D(0.51f, 0.30f, 1f)
            kp[Coco17.LEFT_HIP] = Keypoint2D(0.49f, 0.55f, 1f)
            kp[Coco17.RIGHT_HIP] = Keypoint2D(0.51f, 0.55f, 1f)
            kp[Coco17.RIGHT_WRIST] = Keypoint2D(wx, 0.5f, 1f)
            PoseFrame2D(i, i * 100L, kp)
        }

    private fun stroke(start: Int, peak: Int, end: Int) =
        Stroke2D(strokeIndex = 0, startFrame = start, peakFrame = peak, endFrame = end, peakSpeed = 2.4f)

    /**
     * Builds a session of strokes with controlled dx sign + peakSpeed: stroke i owns
     * frames [3i, 3i+2], the wrist sweeping 0.5 → 0.5 + sign·0.1 → 0.5 so that
     * dx(start→peak) carries exactly the requested sign.
     */
    private fun session(
        specs: List<Pair<Float, Float>>, // (dxSign, peakSpeed)
        noseX: Float
    ): Pair<List<PoseFrame2D>, List<Stroke2D>> {
        val xs = mutableListOf<Float>()
        val strokes = mutableListOf<Stroke2D>()
        specs.forEachIndexed { i, (sign, speed) ->
            val base = 3 * i
            xs += listOf(0.5f, 0.5f + sign * 0.1f, 0.5f)
            strokes += Stroke2D(
                strokeIndex = i, startFrame = base, peakFrame = base + 1,
                endFrame = base + 2, peakSpeed = speed
            )
        }
        return frames(xs, noseX) to strokes
    }

    @Test
    fun keepsForwardStrokeDropsRecoverySwing() {
        // wrist sweeps +x (forward, facing +x) then back -x (recovery)
        val xs = listOf(0.50f, 0.55f, 0.62f, 0.70f, 0.72f, 0.68f, 0.60f, 0.53f, 0.50f)
        val f = frames(xs)
        val forward = stroke(start = 0, peak = 3, end = 4)
        val recovery = stroke(start = 4, peak = 7, end = 8)
        val kept = ForwardStrokeFilter.filter(listOf(forward, recovery), f, Handedness.RIGHT)
        assertEquals(listOf(forward), kept)
    }

    @Test
    fun mirroredPlayerKeepsMirroredForwardStroke() {
        // Player faces -x (nose left of shoulders): forward stroke moves wrist -x
        val xs = listOf(0.50f, 0.45f, 0.38f, 0.30f, 0.28f, 0.32f, 0.40f, 0.47f, 0.50f)
        val f = frames(xs, noseX = 0.45f)
        val forward = stroke(start = 0, peak = 3, end = 4)
        val recovery = stroke(start = 4, peak = 7, end = 8)
        assertEquals(listOf(forward), ForwardStrokeFilter.filter(listOf(forward, recovery), f, Handedness.RIGHT))
    }

    @Test
    fun indeterminateDirectionIsDropped() {
        // Nose dead-centered over shoulder-mid → facing unknown → unverifiable rep
        val xs = listOf(0.50f, 0.55f, 0.62f, 0.70f, 0.72f)
        val f = frames(xs, noseX = 0.50f)
        val s = stroke(start = 0, peak = 3, end = 4)
        assertTrue(ForwardStrokeFilter.filter(listOf(s), f, Handedness.RIGHT).isEmpty())
    }

    @Test
    fun speedDominanceKeepsFastGroupPositiveX() {
        // 3 fast dx>0 strokes (~8f) vs 2 slow dx<0 (~6f) → ratio ~1.33 > 1.2.
        // Head facing CONTRADICTS (nose on the -x side): dominance must override it.
        val (f, strokes) = session(
            specs = listOf(1f to 8.0f, 1f to 8.2f, 1f to 7.9f, -1f to 6.0f, -1f to 6.1f),
            noseX = 0.45f
        )
        val kept = ForwardStrokeFilter.filter(strokes, f, Handedness.RIGHT)
        assertEquals(strokes.take(3), kept)
    }

    @Test
    fun speedDominanceKeepsFastGroupNegativeX() {
        // Mirror polarity: fast group moves -x, head facing contradicts (+x side).
        val (f, strokes) = session(
            specs = listOf(-1f to 8.0f, -1f to 8.2f, -1f to 7.9f, 1f to 6.0f, 1f to 6.1f),
            noseX = 0.55f
        )
        val kept = ForwardStrokeFilter.filter(strokes, f, Handedness.RIGHT)
        assertEquals(strokes.take(3), kept)
    }

    @Test
    fun speedTieFallsBackToHeadFacing() {
        // Near-equal medians (6.55/6.05 ≈ 1.08 < 1.2) → no dominance → head facing
        // (+x) decides and keeps the (slightly slower!) positive group.
        val (f, strokes) = session(
            specs = listOf(1f to 6.0f, 1f to 6.1f, -1f to 6.5f, -1f to 6.6f),
            noseX = 0.55f
        )
        val kept = ForwardStrokeFilter.filter(strokes, f, Handedness.RIGHT)
        assertEquals(strokes.take(2), kept)
    }

    @Test
    fun singleJunkSpikeCannotFlipTheVote() {
        // 4 forward strokes dx>0 @6f plus ONE dx<0 junk spike @9f: minority group
        // size 1 < MIN_GROUP_SIZE → no dominance verdict → fallback to head facing
        // (nose +x) → the 4 forward strokes kept, junk dropped.
        val (f, strokes) = session(
            specs = listOf(1f to 6.0f, 1f to 6.0f, 1f to 6.0f, 1f to 6.0f, -1f to 9.0f),
            noseX = 0.55f
        )
        val kept = ForwardStrokeFilter.filter(strokes, f, Handedness.RIGHT)
        assertEquals(strokes.take(4), kept)
    }

    @Test
    fun gatedWristIsDropped() {
        val xs = listOf(0.50f, 0.55f, 0.62f, 0.70f, 0.72f)
        val f = frames(xs).map { fr ->
            val kp = fr.keypoints.toMutableList()
            kp[Coco17.RIGHT_WRIST] = kp[Coco17.RIGHT_WRIST].copy(score = 0.1f)
            fr.copy(keypoints = kp)
        }
        val s = stroke(start = 0, peak = 3, end = 4)
        assertTrue(ForwardStrokeFilter.filter(listOf(s), f, Handedness.RIGHT).isEmpty())
    }

    @Test
    fun continuousPlayBledStartBoundaryStillReadsForward() {
        // L-28: on continuous play the boundary walk bleeds startFrame back into
        // the previous follow-through (x already forward of the backswing trough),
        // so start→peak displacement reads BACKWARD on a true forward drive.
        // The ~100 ms approach INTO the peak must read forward instead.
        // x: follow-through 0.72 → backswing trough 0.50 → drive up to 0.71
        val xs = listOf(0.72f, 0.60f, 0.50f, 0.55f, 0.65f, 0.70f, 0.71f)
        val f = frames(xs) // nose +x (facing forward = +x)
        val drive = stroke(start = 0, peak = 4, end = 6) // x[4]−x[0] = −0.07 < 0
        val kept = ForwardStrokeFilter.filter(listOf(drive), f, Handedness.RIGHT)
        assertEquals(listOf(drive), kept)
    }
}

File: poses_viewer/src/drill2d/__tests__/forwardStrokeFilter.test.ts
import { describe, expect, it } from 'vitest'
import { filterForwardStrokes } from '../forwardStrokeFilter'
import { Coco17, Keypoint2D, PoseFrame2D, Stroke2D } from '../types'

/** Mirrors ForwardStrokeFilterTest.kt. Player faces +x: nose ahead of shoulder-mid. */
function frames(wristXs: number[], noseX = 0.55): PoseFrame2D[] {
  return wristXs.map((wx, i) => {
    const kp: Keypoint2D[] = Array.from({ length: 17 }, () => ({ x: 0.5, y: 0.5, score: 1 }))
    kp[Coco17.NOSE] = { x: noseX, y: 0.15, score: 1 }
    kp[Coco17.LEFT_SHOULDER] = { x: 0.49, y: 0.30, score: 1 }
    kp[Coco17.RIGHT_SHOULDER] = { x: 0.51, y: 0.30, score: 1 }
    kp[Coco17.LEFT_HIP] = { x: 0.49, y: 0.55, score: 1 }
    kp[Coco17.RIGHT_HIP] = { x: 0.51, y: 0.55, score: 1 }
    kp[Coco17.RIGHT_WRIST] = { x: wx, y: 0.5, score: 1 }
    return { frameIndex: i, timestampMs: i * 100, keypoints: kp }
  })
}

const stroke = (start: number, peak: number, end: number): Stroke2D =>
  ({ strokeIndex: 0, startFrame: start, peakFrame: peak, endFrame: end, peakSpeed: 2.4 })

/**
 * Session of strokes with controlled dx sign + peakSpeed: stroke i owns frames
 * [3i, 3i+2], wrist sweeping 0.5 → 0.5 + sign·0.1 → 0.5.
 */
function session(specs: Array<[number, number]>, noseX: number): [PoseFrame2D[], Stroke2D[]] {
  const xs: number[] = []
  const strokes: Stroke2D[] = []
  specs.forEach(([sign, speed], i) => {
    const base = 3 * i
    xs.push(0.5, 0.5 + sign * 0.1, 0.5)
    strokes.push({ strokeIndex: i, startFrame: base, peakFrame: base + 1, endFrame: base + 2, peakSpeed: speed })
  })
  return [frames(xs, noseX), strokes]
}

describe('filterForwardStrokes (mirrors ForwardStrokeFilterTest)', () => {
  it('keeps forward stroke, drops recovery swing', () => {
    const xs = [0.50, 0.55, 0.62, 0.70, 0.72, 0.68, 0.60, 0.53, 0.50]
    const f = frames(xs)
    const forward = stroke(0, 3, 4)
    const recovery = stroke(4, 7, 8)
    expect(filterForwardStrokes([forward, recovery], f, 'right')).toEqual([forward])
  })

  it('mirrored player keeps mirrored forward stroke', () => {
    const xs = [0.50, 0.45, 0.38, 0.30, 0.28, 0.32, 0.40, 0.47, 0.50]
    const f = frames(xs, 0.45)
    const forward = stroke(0, 3, 4)
    const recovery = stroke(4, 7, 8)
    expect(filterForwardStrokes([forward, recovery], f, 'right')).toEqual([forward])
  })

  it('indeterminate direction is dropped', () => {
    const xs = [0.50, 0.55, 0.62, 0.70, 0.72]
    const f = frames(xs, 0.50) // nose dead-centered → facing unknown
    expect(filterForwardStrokes([stroke(0, 3, 4)], f, 'right')).toEqual([])
  })

  it('speed dominance keeps the fast group (+x), overriding contradicting head facing', () => {
    const [f, strokes] = session([[1, 8.0], [1, 8.2], [1, 7.9], [-1, 6.0], [-1, 6.1]], 0.45)
    expect(filterForwardStrokes(strokes, f, 'right')).toEqual(strokes.slice(0, 3))
  })

  it('speed dominance keeps the fast group (−x), overriding contradicting head facing', () => {
    const [f, strokes] = session([[-1, 8.0], [-1, 8.2], [-1, 7.9], [1, 6.0], [1, 6.1]], 0.55)
    expect(filterForwardStrokes(strokes, f, 'right')).toEqual(strokes.slice(0, 3))
  })

  it('speed tie falls back to head facing', () => {
    const [f, strokes] = session([[1, 6.0], [1, 6.1], [-1, 6.5], [-1, 6.6]], 0.55)
    expect(filterForwardStrokes(strokes, f, 'right')).toEqual(strokes.slice(0, 2))
  })

  it('a single junk spike cannot flip the vote', () => {
    const [f, strokes] = session([[1, 6.0], [1, 6.0], [1, 6.0], [1, 6.0], [-1, 9.0]], 0.55)
    expect(filterForwardStrokes(strokes, f, 'right')).toEqual(strokes.slice(0, 4))
  })

  it('gated wrist is dropped', () => {
    const xs = [0.50, 0.55, 0.62, 0.70, 0.72]
    const f = frames(xs).map(fr => ({
      ...fr,
      keypoints: fr.keypoints.map((k, j) => (j === Coco17.RIGHT_WRIST ? { ...k, score: 0.1 } : k)),
    }))
    expect(filterForwardStrokes([stroke(0, 3, 4)], f, 'right')).toEqual([])
  })
})
